src/main/kotlin/org/jitsi/nlj/rtp/bandwidthestimation/BandwidthEstimator.kt

```diff
@@ -0,0 +1,95 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.rtp.bandwidthestimation
+
+import org.jitsi.nlj.stats.NodeStatsBlock
+import java.time.Instant
+
+/**
+ * An abstract interface to a bandwidth estimation algorithm.
+ *
+ * The invoker of the algorithm will periodically call [processPacketArrival]
+ * and/or [processPacketLoss] as it learns information about packets
+ * that have traversed the network.
+ *
+ * All bandwidths/bitrates are in bits per second.
+ */
+interface BandwidthEstimator {
+    /** The name of the algorithm implemented by this [BandwidthEstimator]. */
+    val algorithmName : String
+
+    /** The initial bandwidth estimate. */
+    var initBw: Float
+
+    /** The minimum bandwidth the estimator will return. */
+    var minBw: Float
+
+    /** The maximum bandwidth the estimator will return. */
+    var maxBw: Float
+
+    /**
+     * Inform the bandwidth estimator about a packet that has arrived at its
+     * destination.
+     *
+     * This function will be called at most once for any value of [seq];
+     * however, it may be called after a call to [processPacketLoss] for the
+     * same [seq] value, if a packet is delayed.
+     *
+     * It is possible (e.g., if feedback was lost) that neither
+     * [processPacketArrival] nor [processPacketLoss] is called for a given [seq].
+     *
+     * The clocks reported by [now], [sendTime], and [recvTime] do not
+     * necessarily have any relationship to each other, but must be consistent
+     * within themselves across all calls to functions of this [BandwidthEstimator].
+     *
+     * @param[now] The current time, when this function is called.
+     * @param[sendTime] The time the packet was sent, if known, or null.
+     * @param[recvTime] The time the packet was received, if known, or null.
+     * @param[seq] A 16-bit sequence number of packets processed by this
+     *  [BandwidthEstimator].
+     * @param[size] The size of the packet, in bytes.
+     * @param[ecn] The ECN markings with which the packet was received.
+     */
+    fun processPacketArrival(now: Instant,
+                             sendTime: Instant?,
+                             recvTime: Instant?,
+                             seq: Int,
+                             size: Int,
+                             ecn: Byte = 0)
+
+    /**
+     * Inform the bandwidth estimator that a packet was lost.
+     *
+     * @param[now] The current time, when this function is called.
+     * @param[sendTime] The time the packet was sent, if known, or null.
+     * @param[seq] A 16-bit sequence number of packets processed by this
+     *  [BandwidthEstimator].
+     */
+    fun processPacketLoss(now: Instant, sendTime: Instant?, seq: Int)
```

Comment:
This makes sense for the send-side bandwidth estimation scheme, but it doesn't seem to fit well with REMB. There the only option is to feed loss based from RRs (right?).

How do you see this working with REMB -- are we going to extend the interface with something like `processRemb` and `processLoss`, or is there another approach you have in mind?

Reply:
I'm still thinking about how I want to handle the send-side (loss-based) part of REMB handling, especially since one of my motivations for this had been to decouple congestion control from RTCP packet loss stats (so we can simplify packet routing by skipping downstream sequence numbers if we have upstream loss).

I may just continue to treat that module as separate from this abstraction, for now.